feature : DB was migrated from sqlite to mysql

waqarRns opened the pull request:
feature : Get latest block api was added
feature : Get latest transactions api was added
feature : Get block summary api was added
feature : Get blocks transaction api was added
feature : Get blocks enrollment api was added

waqarRns committed:
DB was migrated from sqlite to mysql

feature : Get latest block api was added

feature : Get latest transactions api was added

feature : Get block summary api was added

feature : Get blocks transaction api was added

feature : Get blocks enrollment api was added

kchulj commented:
Why was `commitment` renamed to `random_seed`?

TrustHenry commented:
> Why was `commitment` renamed to `random_seed`?

`Git rebase` is required with the latest commit.

TrustHenry commented on src/Stoa.ts:
```diff
@@ -761,11 +768,14 @@ class Stoa extends WebService
                 }
 
                 let overview: ITxOverview = {
+                    status: 'Confirmed',
                     height: JSBI.BigInt(data.tx[0].height).toString(),
                     time: data.tx[0].block_time,
                     tx_hash: new Hash(data.tx[0].tx_hash, Endian.Little).toString(),
                     tx_type: ConvertTypes.TxTypeToString(data.tx[0].type),
+                    tx_size: data.tx[0].tx_size,
                     unlock_height: JSBI.BigInt(data.tx[0].unlock_height).toString(),
+                    lock_height: JSBI.BigInt(data.tx[0].lock_height).toString(),
```
This part is not related to this PR, so it should be separated into a separate PR or a separate commit.

TrustHenry commented on src/Stoa.ts:
```diff
@@ -730,18 +737,18 @@ class Stoa extends WebService
+        let txHash: string = String(req.params.hash);
 
-        logger.http(`GET /wallet/transaction/overview/${hash}}`);
+        logger.http(`GET /wallet/transaction/overview/${txHash}}`);
 
         let tx_hash: Hash;
         try
         {
-            tx_hash = new Hash(hash);
+            tx_hash = new Hash(txHash);
         }
         catch (error)
         {
-            res.status(400).send(`Invalid value for parameter 'hash': ${hash}`);
+            res.status(400).send(`Invalid value for parameter 'hash': ${txHash}`);
```
Comments as below

waqarRns replied:
here we are using  req.params, for invalid hash it will be catched by catch block. isn't it ok?

TrustHenry commented on src/Stoa.ts:
```diff
@@ -774,21 +784,318 @@ class Stoa extends WebService
+        else if (req.query.hash !== undefined) {
+            field = 'hash';
+            const hash: string = String(req.query.hash);
+            value = new Hash(hash).toBinary(Endian.Little);
```
Error handling is required.

waqarRns replied:
` if (req.query.height !== undefined && Utils.isPositiveInteger(req.query.height.toString())) {
            field = 'height';
            value = String(req.query.height);
        }
         // Validating Parameter - hash
        else if (req.query.hash !== undefined) {
            field = 'hash';
            const hash: string = String(req.query.hash);
            value = new Hash(hash).toBinary(Endian.Little);
        }
        else {
            res.status(400).send(`Invalid value for parameter 'height': ${req.query.height} and 'hash': ${req.query.hash}`);
            return;
        }`

waqarRns replied:
`else {
            res.status(400).send(`Invalid value for parameter 'height': ${req.query.height} and 'hash': ${req.query.hash}`);
            return;
        }`

waqarRns replied:
Error handle is there please review again.

TrustHenry replied:
https://github.com/bosagora/boa-sdk-ts/blob/ffeb2d96c73d6d465862266fcc6fe3a08c7a02af/src/modules/common/Hash.ts#L51-L52

waqarRns replied:
i fixed this

TrustHenry commented on src/Stoa.ts:
```diff
@@ -774,21 +784,318 @@ class Stoa extends WebService
+                    signature:  new Hash(data[0].signature, Endian.Little).toString(),
+                    random_seed:  new Hash(data[0].random_seed, Endian.Little).toString(),
+                    time: data[0].time_stamp,
+                    version: "v0.x.x",
```
Better read it from the settings.

waqarRns replied:
i will update this in upcoming PR.

TrustHenry commented on src/Stoa.ts:
```diff
@@ -774,21 +784,318 @@ class Stoa extends WebService
+        else if (req.query.hash !== undefined) {
+            field = 'hash';
+            const hash: string = String(req.query.hash);
+            value = new Hash(hash).toBinary(Endian.Little);
```
Error handling is required.

waqarRns replied:
```suggestion
            
```

waqarRns replied:
```please review
             if (req.query.height !== undefined && Utils.isPositiveInteger(req.query.height.toString())) {
            field = 'height';
            value = String(req.query.height);
        }
        // Validating Parameter - hash
        else if (req.query.hash !== undefined) {
            field = 'hash';
            const hash: string = String(req.query.hash);
            value = new Hash(hash).toBinary(Endian.Little);
        }
        else {
            res.status(400).send(`Invalid value for parameter 'height': ${req.query.height} and 'hash': ${req.query.hash}`);
            return;
        }
```

waqarRns replied:
if both are undefined then it will else condition will be executed and if height or hash is defined the code will proceed

TrustHenry replied:
https://github.com/bosagora/boa-sdk-ts/blob/ffeb2d96c73d6d465862266fcc6fe3a08c7a02af/src/modules/common/Hash.ts#L85
If the length verification fails, an error is thrown.

waqarRns replied:
i fixed this

TrustHenry commented on src/Stoa.ts:
```diff
@@ -774,21 +784,318 @@ class Stoa extends WebService
+                return;
+            }
+            limit = Number(req.query.limit.toString());
+            if (limit > 10) {
+                res.status(400).send(`Limit cannot be a number greater than 10: ${limit}`);
+                return;
+            }
+        }
+        else
+            limit = 10;
```
Please use the `limit` as a constant.
The same is true of other codes.

waqarRns replied:
as limit is page_size stoa is using it as ` let page_size: number; ` we are keeping the things similier to stoa

TrustHenry replied:
The page size of BOASCAN can be set at 10, 25, 50, 100.

TrustHenry commented on src/Stoa.ts:
```diff
@@ -774,21 +784,318 @@ class Stoa extends WebService
+                let enrollmentElementList: Array<IBlockEnrollmentElements> = [];
+                for (const row of data.enrollments) {
+                    enrollmentElementList.push(
+    {
```
```suggestion
                                  {
```

waqarRns replied:
updated

TrustHenry commented on src/Stoa.ts:
```diff
@@ -774,21 +784,318 @@ class Stoa extends WebService
+        else if (req.query.hash !== undefined) {
+            field = 'hash';
+            const hash: string = String(req.query.hash);
+            value = new Hash(hash).toBinary(Endian.Little);
```
Error handling is required.

waqarRns replied:
`  if (req.query.height !== undefined && Utils.isPositiveInteger(req.query.height.toString())) {
            field = 'height';
            value = String(req.query.height);
        }
        // Validating Parameter - hash
        else if (req.query.hash !== undefined) {
            field = 'hash';
            const hash: string = String(req.query.hash);
            value = new Hash(hash).toBinary(Endian.Little);
        }
        else {
            res.status(400).send(`Invalid value for parameter 'height': ${req.query.height} and 'hash': ${req.query.hash}`);
            return;
        }`

waqarRns replied:
error handling is there please review else block.

TrustHenry commented on src/Stoa.ts:
```diff
@@ -774,21 +784,318 @@ class Stoa extends WebService
+                    height: data[0].height,
+                    transactions: data[0].transactions,
+                    validators: data[0].validators,
+                    frozen_coin: 5283595, //static data because of unavailability of real data
+                    circulating_supply: 5283535,
+                    active_validators: 155055
```
//FIX ME static data because of unavailability of real data

waqarRns replied:
updated

TrustHenry reviewed (commented):
Let's stop here.
I'll keep reviewing it.

The unit test must be passed.
Please add mysql environment in Github actions CI.

TrustHenry commented on .github/workflows/main.yml:
```diff
@@ -11,15 +8,31 @@ jobs:
-    - run: npm test
+    - run:  CI=true npm run
+      env:
+        CI: true
```
```suggestion
    steps:
    - uses: actions/checkout@v2
    - name: Shutdown Ubuntu MySQL (SUDO)
      run: sudo service mysql stop
    - name: Use Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}
    - name: Set up MySQL 5.7
      uses: mirromutth/mysql-action@master
      with:
        mysql version: 5.7
        mysql database: root
        mysql root password: root
    - name: npm install, build, and test
      run: |
        npm install
        npm run build --if-present
        npm test
      env:
        CI: true
```

waqarRns replied:
updated

TrustHenry commented on .github/workflows/main.yml:
```diff
@@ -11,15 +8,31 @@ jobs:
       uses: actions/setup-node@v1
       with:
         node-version: ${{ matrix.node-version }}
+    - run: |
+           sudo service mysql start
+    - run: mysql -h 127.0.0.1 --port 8888 -u root -ppassword -e 'SHOW DATABASES;'
+    - run: mysql -h 127.0.0.1 --port 8888 -u root -ppassword -e 'SET GLOBAL sql_mode=(SELECT REPLACE(@@sql_mode,"ONLY_FULL_GROUP_BY",""));'
```
```suggestion
    - run: |
           sudo service mysql start
           mysql -h 127.0.0.1 --port 8888 -u root -ppassword -e 'SHOW DATABASES;'
           mysql -h 127.0.0.1 --port 8888 -u root -ppassword -e 'SET GLOBAL sql_mode=(SELECT REPLACE(@@sql_mode,"ONLY_FULL_GROUP_BY",""));'
```

waqarRns replied:
resolved

TrustHenry commented on .github/workflows/main.yml:
```diff
@@ -11,15 +8,31 @@ jobs:
+    - run: mysql -h 127.0.0.1 --port 8888 -u root -ppassword -e 'SHOW DATABASES;'
+    - run: mysql -h 127.0.0.1 --port 8888 -u root -ppassword -e 'SET GLOBAL sql_mode=(SELECT REPLACE(@@sql_mode,"ONLY_FULL_GROUP_BY",""));'
     - run: npm install
     - run: npm run build --if-present
-    - run: npm test
+    - run:  CI=true npm run
+      env:
+        CI: true
```
npm install
              npm run build --if-present
              npm test
      env:
        CI: true

waqarRns replied:
updated

TrustHenry commented on .github/workflows/main.yml:
```diff
@@ -11,15 +8,31 @@ jobs:
       uses: actions/setup-node@v1
       with:
         node-version: ${{ matrix.node-version }}
+    - run: |
```
```suggestion
    - name: Setup mysql, npm install, build, and test
    - run: |
```

waqarRns replied:
updated

TrustHenry commented on tests/Stoa.test.ts:
```diff
@@ -54,7 +57,8 @@ describe ('Test of Stoa API Server', () =>
     {
-        stoa_server = new TestStoa(new URL("http://127.0.0.1:2826"), port);
+        testDBConfig = await MockDBConfig();
+        stoa_server = new TestStoa(testDBConfig,new URL("http://127.0.0.1:2826"), port);
```
```suggestion
        stoa_server = new TestStoa(testDBConfig, new URL("http://127.0.0.1:2826"), port);
```

waqarRns replied:
updated

TrustHenry commented on tests/Stoa.test.ts:
```diff
@@ -206,9 +212,9 @@ describe ('Test of Stoa API Server', () =>
                 new Hash("0x6100ee7a7e00e18e06b743a7ae90e91781c09e0f1791ee2849ce15caf4c6ee1f3aebc23768f98153d8e3fb10ac66267e06acc31dccbfdbe671294a7fded22432");
-            const commitment =
+            const random_seed =
                 new Hash("0xe0c04a5bd47ffc5b065b7d397e251016310c43dc77220bf803b73f1183da00b0e67602b1f95cb18a0059aa1cdf2f9adafe979998364b38cd5c15d92b9b8fd815");
-            const enrollment = new Enrollment(utxo_key, commitment, 20, enroll_sig);
+            const enrollment = new Enrollment(utxo_key, random_seed, 20, enroll_sig);
```
Please don't change it.

waqarRns replied:
this has been changed

TrustHenry commented on tests/Stoa.test.ts:
```diff
@@ -632,6 +639,470 @@ describe ('Test of Stoa API Server', () =>
         };
         assert.deepStrictEqual(response.data, expected);
     });
+    it('Test of the path /latest-blocks', async () => {
```
Could you separate it into a separate test file for easy brunch merge later?

waqarRns replied:
i will update this in new PR.

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -128,7 +129,7 @@ export class LedgerStorage extends Storages
             tx_hash             BLOB    NOT NULL,
             utxo_key            BLOB    NOT NULL,
-            amount              NUMERIC NOT NULL,
+            amount              DECIMAL(65) NOT NULL,
```
```suggestion
            amount              BIGINT(20) UNSIGNED NOT NULL,
```

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -118,7 +119,7 @@ export class LedgerStorage extends Storages
             utxo                BLOB    NOT NULL,
```
```suggestion
            utxo                TINYBLOB    NOT NULL,
```

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -118,7 +119,7 @@ export class LedgerStorage extends Storages
             unlock_bytes        BLOB    NOT NULL,
             unlock_age          INTEGER NOT NULL,
-            PRIMARY KEY(block_height, tx_index, in_index, utxo)
+            PRIMARY KEY(block_height, tx_index, in_index, utxo(255))
```
```suggestion
            PRIMARY KEY(block_height, tx_index, in_index, utxo(64))
```

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -190,7 +191,7 @@ export class LedgerStorage extends Storages
             payload_fee         INTEGER NOT NULL,
             tx_size             INTEGER NOT NULL,
-            PRIMARY KEY(tx_hash)
+            PRIMARY KEY(tx_hash(255))
```
```suggestion
            PRIMARY KEY(tx_hash(64))
```

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -141,28 +142,28 @@ export class LedgerStorage extends Storages
             tx_hash             BLOB    NOT NULL,
```
```suggestion
            tx_hash             TINYBLOB    NOT NULL,
```

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -128,7 +129,7 @@ export class LedgerStorage extends Storages
             output_index        INTEGER NOT NULL,
             tx_hash             BLOB    NOT NULL,
             utxo_key            BLOB    NOT NULL,
```
```suggestion
            tx_hash             TINYBLOB    NOT NULL,
            utxo_key            TINYBLOB    NOT NULL,
```

TrustHenry reviewed (commented):
Set the Hash data field Mysql type to TINYBLOB and specify 64 bytes as the key.

waqarRns committed:
Feature: Error handling was added
Feature: DB schema was updated(BLOB=>TINYBLOB, DECIMAL(65)=>BIGINT(20))
Feature: Testcases were updated

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -1551,7 +1634,11 @@ export class LedgerStorage extends Storages
                 S.address,
                 S.amount,
-                S.utxo_key as utxo
+                S.utxo_key as utxo,
+				B.signature,
+				I.in_index,
+				I.unlock_age,
+				I.unlock_bytes as bytes
```
```suggestion
                S.utxo_key as utxo,
		B.signature,
		I.in_index,
		I.unlock_age,
		I.unlock_bytes as bytes
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -1560,9 +1647,12 @@ export class LedgerStorage extends Storages
+                O.output_index,
                 O.amount,
-                O.utxo_key as utxo
+				O.lock_type,
+				O.lock_bytes as bytes,
```
```suggestion
		O.lock_type,
		O.lock_bytes as bytes,
```

waqarRns replied:
resolved

TrustHenry replied:
Not resolved yet

waqarRns replied:
![codeformat](https://user-images.githubusercontent.com/78407888/117402659-42d4f900-af20-11eb-99f4-06e1935928da.png)

waqarRns replied:
i resolved this.

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -1827,4 +1917,175 @@ export class LedgerStorage extends Storages
+            `SELECT
+                height, hash , merkle_root , signature , validators , tx_count, enrollment_count ,time_stamp 
+            FROM
+                blocks ORDER BY height DESC LIMIT ? OFFSET ?`
```
```suggestion
            `SELECT
                height, hash, merkle_root, signature, validators, tx_count, enrollment_count, time_stamp 
            FROM
                blocks 
            ORDER BY height DESC 
            LIMIT ? OFFSET ?`
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -1827,4 +1917,175 @@ export class LedgerStorage extends Storages
+             ON 
+             ( T.tx_hash = O.tx_hash AND T.block_height = O.block_height ) AND B.height = T.block_height 
+             GROUP BY O.tx_hash ORDER BY T.block_height DESC LIMIT ? OFFSET ?;
+            `
```
```suggestion
        let sql =
            `SELECT 
                T.block_height, T.tx_hash, T.type, T.tx_fee, T.tx_size, Sum(IFNULL(O.amount,0)) as amount, B.time_stamp 
             FROM
                 tx_outputs O 
                 INNER JOIN transactions T ON (T.tx_hash = O.tx_hash)
                 INNER JOIN blocks B ON (B.height = T.block_height) 
             GROUP BY O.tx_hash 
             ORDER BY T.block_height DESC
             LIMIT ? OFFSET ?;`
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -1827,4 +1917,175 @@ export class LedgerStorage extends Storages
+        let countsql =
+                `select count(*) as total_records from  
+                blocks b INNER JOIN enrollments e ON
+                e.block_height = b.height AND b.${field} = ?`;
```
```suggestion
        let sql =
            `SELECT
                E.block_height, E.utxo_key, E.commitment, E.cycle_length, E.enroll_sig 
            FROM
                blocks B
                INNER JOIN enrollments E ON (E.block_height = B.height)
                AND B.${field} = ?
            ORDER BY O.enrollment_index ASC
            LIMIT ? OFFSET ?;`;
        let countsql =
                `SELECT IFNULL(count(*),0) as total_records
                FROM  
                    blocks B 
                    INNER JOIN enrollments E ON (E.block_height = B.height)
                    AND B.${field} = ?`;
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -1827,4 +1917,175 @@ export class LedgerStorage extends Storages
+            result.enrollments = rows;
+            return this.query(countsql,[value]);
+        })
+        .then((rows :any [])=>{
```
```suggestion
        .then((rows : any[])=>{
            result.enrollments = rows;
            return this.query(countsql,[value]);
        })
        .then((rows : any[])=>{
```

waqarRns replied:
resolved

waqarRns committed:
Feature : paginate method was added

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -1827,4 +1917,175 @@ export class LedgerStorage extends Storages
+     * and if an error occurs the `.catch` is called with an error. 
+     */
+    public getBlockTransactions(field: string, value: string | Buffer, limit: number, page: number): Promise<any[]> {
+        let sql_tx =
+            `SELECT 
+                T.block_height, T.tx_hash, SUM(O.amount) as amount, type,
+                tx_fee, tx_size, B.time_stamp,
+                JSON_ARRAYAGG(JSON_OBJECT("address", O.address, "amount", O.amount)) as receiver,
+            (SELECT
+                    s.address
+                FROM
+                    blocks b
+                    INNER JOIN transactions t ON b.height = t.block_height
+                    INNER JOIN tx_inputs i ON t.tx_hash = i.tx_hash
+                    INNER JOIN tx_outputs s ON i.utxo = s.utxo_key
+                WHERE
+                    ${field} = ? ) as sender_address
+            FROM    
+                tx_outputs O INNER JOIN transactions T  ON ( T.tx_hash = O.tx_hash AND T.block_height = O.block_height) 
+                INNER JOIN blocks B ON  B.height = T.block_height
+            WHERE
+                ${field} = ?
+            GROUP BY T.tx_hash
+            LIMIT ? OFFSET ?;`;
```
```suggestion
        let sql_tx =
            `SELECT 
                T.block_height, T.tx_hash, SUM(IFNULL(O.amount,0)) as amount, T.type,
                T.tx_fee, T.tx_size, B.time_stamp,
                JSON_ARRAYAGG(JSON_OBJECT("address", O.address, "amount", O.amount)) as receiver,
                (SELECT
                    S.address
                FROM
                    blocks B
                    INNER JOIN transactions T ON (B.height = T.block_height)
                    INNER JOIN tx_inputs I ON (T.tx_hash = I.tx_hash)
                    INNER JOIN tx_outputs s ON (i.utxo = s.utxo_key)
                WHERE
                    B.${field} = ? ) as sender_address
            FROM    
                tx_outputs O 
                INNER JOIN transactions T ON (T.tx_hash = O.tx_hash) 
                INNER JOIN blocks B ON  (B.height = T.block_height)
            WHERE
                B.${field} = ?
            GROUP BY T.tx_hash
            ORDER BY T.tx_index ASC
            LIMIT ? OFFSET ?;`;
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -1827,4 +1917,175 @@ export class LedgerStorage extends Storages
+            GROUP BY T.tx_hash
+            LIMIT ? OFFSET ?;`;
+        
+        let sql_count =
+               `SELECT
+                    count(*) as total_records
+                FROM
+                    transactions t INNER JOIN blocks b ON b.height = t.block_height
+                WHERE
+                    ${field} = ?;`;
```
```suggestion
        let sql_count =
               `SELECT
                    IFNULL(count(*),0) as total_records
                FROM
                    transactions T 
                    INNER JOIN blocks B ON (B.height = T.block_height)
                WHERE
                    ${field} = ?;`;
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/common/Config.ts:
```diff
@@ -196,19 +196,48 @@ export class ServerConfig implements IServerConfig
+    /**
+     * The host database port
+     */
+     port: number
+    /** 
+     * multiple Statements exec config
+    */
+    multipleStatements: boolean
```
```suggestion
    port: number
    /** 
     * multiple Statements exec config
    */
    multipleStatements: boolean
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/common/Config.ts:
```diff
@@ -196,19 +196,48 @@ export class ServerConfig implements IServerConfig
+    /**
+     * The host database port
+     */
+     port: number
+    /** 
+     * multiple Statements exec config
+    */
+    multipleStatements: boolean
```
```suggestion
    port: number
    /** 
     * multiple Statements exec config
    */
    multipleStatements: boolean
```

waqarRns replied:
resolved

TrustHenry commented on src/Types.ts:
```diff
@@ -362,3 +647,101 @@ export interface ISPVStatus
+    /**
+     * page size 
+     */
+    pageSize:number,
+    
+    /**
+     * page number
+     */
+     page :number
```
```suggestion
    pageSize:number,
    
    /**
     * page number
     */
    page :number
```

waqarRns replied:
resolved

TrustHenry replied:
Not yet been resolved.

waqarRns replied:
resolved

TrustHenry commented on src/Types.ts:
```diff
@@ -159,11 +159,231 @@ export interface ITxHistoryElement
+    /**
+     * The hash of UTXO
+     */
+    utxo: string;
+    
+    /**
+     * Random seed
+     */
+     commitment: string,
```
```suggestion
    utxo: string;
    
    /**
     * Random seed
     */
    commitment: string,
```

waqarRns replied:
resolved

TrustHenry replied:
Not yet been resolved.

waqarRns replied:
resolved

TrustHenry commented on src/Types.ts:
```diff
@@ -159,11 +159,231 @@ export interface ITxHistoryElement
+    /**
+     * total amount sent in block
+     */
+    total_sent : string;
+    /**
+     * total recieved amount
+     */
+    total_recieved :string;
+    /**
+     * total rewards
+     */
+    total_reward :string
+    /**
+     * total fee for the block
+     */
+    total_fee :string
+    /**
+     * total size of the block
+     */
+    total_size : string
+    /**
+     * Agora version
+     */
+    version : string
```
```suggestion
    total_sent: string;
    /**
     * total recieved amount
     */
    total_recieved: string;
    /**
     * total rewards
     */
    total_reward: string
    /**
     * total fee for the block
     */
    total_fee: string
    /**
     * total size of the block
     */
    total_size: string
    /**
     * Agora version
     */
    version: string
```

waqarRns replied:
resolved

TrustHenry commented on src/Types.ts:
```diff
@@ -159,11 +159,231 @@ export interface ITxHistoryElement
+    /**
+     * total size of the block
+     */
+    total_size : string
```
Why is it defined as string?

waqarRns replied:
resolved

waqarRns committed:
Update : Code format was updated

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -1827,4 +1917,186 @@ export class LedgerStorage extends Storages
+    public getLatestBlocks(limit: number, page: number): Promise<any[]> {
+        let sql =
+            `SELECT
+                height, hash , merkle_root , signature , validators , tx_count,
+                enrollment_count ,time_stamp 
```
```suggestion
                height, hash, merkle_root, signature, validators, tx_count,
                enrollment_count, time_stamp 
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -1827,4 +1917,186 @@ export class LedgerStorage extends Storages
+    public getLatestTransactions(limit: number, page: number): Promise<any[]> {
+        let sql =
+            `SELECT 
+             T.block_height, T.tx_hash, type, tx_fee, tx_size, 
```
```suggestion
             T.block_height, T.tx_hash, T.type, T.tx_fee, T.tx_size, 
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/storage/Storages.ts:
```diff
@@ -13,48 +13,55 @@
                     });
+            }).catch((err)=>{
+                if (callback != null)
+                callback(err);
```
```suggestion
                    callback(err);
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/common/Config.ts:
```diff
@@ -196,19 +196,54 @@ export class ServerConfig implements IServerConfig
+    /** 
+     * multiple Statements exec config
+    */
+
```
```suggestion
```

waqarRns replied:
resolved

TrustHenry commented on src/Types.ts:
```diff
@@ -159,11 +159,237 @@ export interface ITxHistoryElement
+    /**
+     * total fee for the block
+     */
+    total_fee :string
```
```suggestion
    total_fee: string
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -497,6 +513,71 @@ export class LedgerStorage extends Storages
+                  let total_recieved_sql = `select SUM(O.amount) as total_recieved from 
+                  tx_outputs O Inner join blocks B ON (O.block_height = B.height)
+                  where block_height = ?;`;
+                  let transaction_stats = `select  SUM(T.tx_fee) as tx_fee, SUM(T.payload_fee) as payload_fee, SUM(T.tx_size) as total_size 
+                  from 
+                  transactions T Inner join blocks B ON (T.block_height = B.height)
+                  where block_height =?;`;
```
```suggestion
                  let transaction_stats = `SELECT  SUM(IFNULL(T.tx_fee,0)) as tx_fee, SUM(IFNULL(T.payload_fee,0)) as payload_fee, SUM(IFNULL(T.tx_size,0)) as total_size 
                  FROM 
                      transactions T Inner join blocks B ON (T.block_height = B.height)
                  WHERE 
                      block_height =?;`;
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -497,6 +513,71 @@ export class LedgerStorage extends Storages
+                  let total_sent = JSBI.BigInt(0);
+                  let total_fee = JSBI.BigInt(0);
+                  let total_size = JSBI.BigInt(0);
+                  let total_recieved_sql = `select SUM(O.amount) as total_recieved from 
+                  tx_outputs O Inner join blocks B ON (O.block_height = B.height)
+                  where block_height = ?;`;
```
```suggestion
                  let total_recieved_sql = `SELECT SUM(IFNULL(O.amount,0)) as total_recieved
                                                            FROM 
                                                                tx_outputs O
                                                                INNER JOIN blocks B ON (O.block_height = B.height)
                                                           WHERE
                                                               block_height = ?;`;
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/storage/LedgerStorage.ts:
```diff
@@ -497,6 +513,71 @@ export class LedgerStorage extends Storages
+     */
+      public putBlockstats (block: Block): Promise<void>
+      {
+          function save_blockstats (storage :LedgerStorage, height :Height, total_sent :JSBI, total_recieved: JSBI, total_size :JSBI, total_fee :JSBI): Promise<void>
```
```suggestion
          function save_blockstats (storage: LedgerStorage, height: Height, total_sent: JSBI, total_recieved: JSBI, total_size: JSBI, total_fee: JSBI): Promise<void>
```

waqarRns replied:
resolved

TrustHenry commented on src/modules/common/Config.ts:
```diff
@@ -196,19 +196,54 @@ export class ServerConfig implements IServerConfig
         let conf = extend(true, {}, DatabaseConfig.defaultValue());
-        extend(true, conf, {filename: filename});
-        this.filename = path.resolve(Utils.getInitCWD(), conf.filename);
+        extend(true, conf, { host: host, user: user, password: password, database: database, port : port, multipleStatements: multipleStatements });
```
```suggestion
        extend(true, conf, { host: host, user: user, password: password, database: database, port: port, multipleStatements: multipleStatements });
```

waqarRns replied:
resolved

TrustHenry commented on tests/Storage.test.ts:
```diff
@@ -219,19 +223,21 @@ describe ('Test ledger storage and inquiry function.', () =>
 describe ('Test for storing block data in the database', () =>
 {
     let ledger_storage: LedgerStorage;
+    let testDBConfig : IDatabaseConfig;
```
```suggestion
    let testDBConfig: IDatabaseConfig;
```

waqarRns replied:
resolved

TrustHenry commented on tests/Storage.test.ts:
```diff
@@ -18,25 +18,29 @@ import { sample_data, sample_data2, sample_preImageInfo } from "./Utils";
 describe ('Test ledger storage and inquiry function.', () =>
 {
     let ledger_storage: LedgerStorage;
+    let testDBConfig : IDatabaseConfig;
```
```suggestion
    let testDBConfig: IDatabaseConfig;
```

waqarRns replied:
resolved

TrustHenry commented on tests/Storage.test.ts:
```diff
@@ -309,6 +312,7 @@ describe ('Test for storing block data in the database', () =>
 describe ('Tests that sending a pre-image', () =>
 {
     let ledger_storage: LedgerStorage;
+    let testDBConfig : IDatabaseConfig;
```
```suggestion
    let testDBConfig: IDatabaseConfig;
```

waqarRns replied:
resolved

TrustHenry commented on tests/Storage.test.ts:
```diff
@@ -376,20 +382,23 @@ describe ('Tests that sending a pre-image', () =>
 describe ('Tests storing transaction pools of a transaction', () =>
 {
     let ledger_storage: LedgerStorage;
+    let testDBConfig : IDatabaseConfig;
```
```suggestion
    let testDBConfig: IDatabaseConfig;
```

TrustHenry commented on tests/WalletAPI.test.ts:
```diff
@@ -25,6 +27,7 @@ describe ('Test of Stoa API for the wallet', () =>
     let stoa_server: TestStoa;
     let agora_server: TestAgora;
     let client = new TestClient();
+    let testDBConfig : IDatabaseConfig;
```
```suggestion
    let testDBConfig: IDatabaseConfig;
```

waqarRns replied:
resolved

TrustHenry commented on src/Stoa.ts:
```diff
@@ -129,6 +134,12 @@ class Stoa extends WebService
+        this.app.get("/block-summary", this.getBlockSummary.bind(this));
+        this.app.get("/block-enrollments", this.getBlockEnrollments.bind(this));
+        this.app.get("/block-transactions", this.getBlockTransactions.bind(this));
+        this.app.get("/boa-stats", this.getBOAStats.bind(this));
```
`/boa-stats` Test code creation is required.

waqarRns replied:
I added the testcase.

waqarRns committed:
Feature: boa-stats testcase was added
Update: Code refactor

TrustHenry reviewed (approved):
Nice work!

TrustHenry commented:
I think a few styles need to be modified, but I'll upload the revised PR.
